alibi_detect/utils/keops/kernels.py

```diff
@@ -0,0 +1,91 @@
+import numpy as np
+from pykeops.torch import LazyTensor
+import torch
+import torch.nn as nn
+from typing import Callable, Optional
+
+
+def sigma_mean(x: LazyTensor, y: LazyTensor, dist: LazyTensor) -> torch.Tensor:
+    """
+    Bandwidth estimation using the mean heuristic.
+
+    Parameters
+    ----------
+    x
+        LazyTensor of instances with dimension [Nx, 1, features].
+    y
+        LazyTensor of instances with dimension [1, Ny, features].
+    dist
+        LazyTensor with dimensions [Nx, Ny], containing the pairwise distances between `x` and `y`.
+
+    Returns
+    -------
+    The computed bandwidth, `sigma`.
+    """
+    n = x.shape[0]
+    if x.shape == y.shape and (dist.min(axis=1) == 0.).all() and \
+            (torch.arange(n) == dist.argmin(axis=1).cpu().view(-1)).all():
+        n_mean = n * (n - 1)
+    else:
+        n_mean = np.prod(dist.shape)
+    sigma = (.5 * dist.sum(1).sum().unsqueeze(-1) / n_mean) ** .5
+    return sigma
```

Comment:
Not sure I follow the logic, come comments would be appreciated!

1. What are the two different branch conditions?
2. Doesn't `x.shape == y.shape` imply that `Nx == Ny ==1` (from docstring shapes)? But then the following don't make sense to me as `dist` would have shape `(1, 1)`.

Reply:
1. Checks if you're comparing x with itself (y == x) to take zero diagonal into account. You need to do it in a roundabout way given we're working on LazyTensors here.
2. Same as 1., if `x!=y` then you go to the else statement. Note that dist is also broadcasted and does not have dims of 1 anymore.

Reply:
Note that it isn't entirely foolproof: if for instance x takes consecutive elements which are the same there might be more 0's in the distance matrix where the argmin comes before the diagonal (the last check). Can avoid this I think by extracting the diagonal from `dist.argmin(axis=1)` and checking `torch.arange(n)` against that.

Reply:
Would be great to add these comments in the code as it simplifies the comprehension a lot!

I'm still not sure about `x.shape == y.shape` because the docstring specifies shapes to be `[Nx, 1, features]` and `[1, Ny, features]`. Does it mean that in the special case that the same tensor is passed in both we actually support shapes`[Nx, 1, features]` and `[Nx, 1, features]`? If so, should this be documented (if it's intentional to make it easier for users to write `sigma_mean(x, x)` rather than reshuffling the axis on the second `x`)?